db/repair_bucket.go

```diff
@@ -0,0 +1,228 @@
+type RepairJobType string
+
+const (
+	RepairRevTreeCycles = RepairJobType("RepairRevTreeCycles")
+)
+
+// Params suitable for external (eg, HTTP) invocations to describe a RepairBucket operation
+type RepairBucketParams struct {
+	DryRun     bool              `json:"dry_run"`
+	RepairJobs []RepairJobParams `json:"repair_jobs"`
+}
+
+// Params suitable for external (eg, HTTP) invocations to describe a specific RepairJob operation
+type RepairJobParams struct {
+	RepairJobType   RepairJobType          `json:"repair_job_type"`
+	RepairJobParams map[string]interface{} `json:"repair_job_params"`
+}
+
+// Given a Couchbase Bucket doc, transform the doc in some way to produce a new doc.
+// Also return a boolean to indicate whether a transformation took place, or any errors occurred.
+type DocTransformer func(doc *document) (transformedDoc *document, transformed bool, err error)
+
+// A RepairBucket struct is the main API entrypoint to call for repairing documents in buckets
+type RepairBucket struct {
+	DryRun                  bool // If true, will only output what changes it *would* have made, but not make any changes
+	WriteRepairedDocsToDisk bool // If true, write the before and after doc to temp files
+	Bucket                  base.Bucket
+	RepairJobs              []DocTransformer
+}
+
+func NewRepairBucket(bucket base.Bucket) *RepairBucket {
+	return &RepairBucket{
+		Bucket: bucket,
+	}
+}
+
+func (r *RepairBucket) SetDryRun(dryRun bool) *RepairBucket {
+	r.DryRun = dryRun
+	return r
+}
+
+func (r *RepairBucket) AddRepairJob(repairJob DocTransformer) *RepairBucket {
+	r.RepairJobs = append(r.RepairJobs, repairJob)
+	return r
+}
+
+func (r *RepairBucket) InitFrom(params RepairBucketParams) *RepairBucket {
+
+	r.SetDryRun(params.DryRun)
+	r.WriteRepairedDocsToDisk = true
+	for _, repairJobParams := range params.RepairJobs {
+		switch repairJobParams.RepairJobType {
+		case RepairRevTreeCycles:
+			r.AddRepairJob(RepairJobRevTreeCycles)
+		}
+	}
+
+	return r
+}
+
+func (r RepairBucket) RepairBucket() (repairedDocs []*document, err error) {
+
+	options := Body{"stale": false, "reduce": false}
+	options["startkey"] = []interface{}{true}
+
+	vres, err := r.Bucket.View(DesignDocSyncHousekeeping, ViewImport, options)
+	if err != nil {
+		return repairedDocs, err
+	}
+
+	for _, row := range vres.Rows {
+		rowKey := row.Key.([]interface{})
+		docid := rowKey[1].(string)
+		key := realDocID(docid)
+
+		err = r.Bucket.Update(key, 0, func(currentValue []byte) ([]byte, error) {
+			// Be careful: this block can be invoked multiple times if there are races!
+			if currentValue == nil {
+				return nil, couchbase.UpdateCancel // someone deleted it?!
+			}
+			doc, err := unmarshalDocument(docid, currentValue)
+			if err != nil {
+				return nil, err
+			}
+			updatedDoc, shouldUpdate, err := r.TransformBucketDoc(doc)
+			if err != nil {
+				return nil, err
+			}
+
+			switch shouldUpdate {
+			case true:
+
+				if r.WriteRepairedDocsToDisk {
```

Comment:
I'm not sure writing to disk is the best approach - if this is a REST call, retrieving the document from disk wherever Sync Gateway is running might not be trivial.  If we're trying to provide a way for people to roll back the change, how about an optional expiring backup of any modified documents in the bucket?  (as _sync:repair:[key])?

Reply:
Fixed